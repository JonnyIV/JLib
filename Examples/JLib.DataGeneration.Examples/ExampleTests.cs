using FluentAssertions;
using JLib.DataGeneration.Examples.DataPackages;
using JLib.DataGeneration.Examples.Setup.Models;
using JLib.DataGeneration.Examples.Setup.SystemUnderTest;
using JLib.DataGeneration.Examples.SnapshotInfo;
using JLib.DependencyInjection;
using JLib.Exceptions;
using JLib.AutoMapper;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Snapshooter;
using Snapshooter.Xunit;
using Xunit;
using Xunit.Abstractions;
using Xunit.Sdk;

namespace JLib.DataGeneration.Examples;

public class ExampleTests : IDisposable
{
    private readonly List<IDisposable> _disposables = new();
    private readonly IServiceProvider _provider;
    private readonly ShoppingServiceMock _shoppingService;
    private readonly IIdRegistry _idRegistry;

    public ExampleTests(ITestOutputHelper testOutputHelper)
    {
        var exceptions = ExceptionBuilder.Create("setup");

        var loggerFactory = new LoggerFactory()
            .AddXunit(testOutputHelper);

        var services = new ServiceCollection()
            // executes and caches reflection on all given types
            .AddTypeCache(out var typeCache, exceptions, loggerFactory, JLibDataGenerationExamplesTypePackage.Instance)
            // add all type packages
            .AddDataPackages(typeCache, new() { DefaultNamespace = "JLib.DataGeneration.Examples.DataPackages" })
            // add all automapper profiles (required profile: JLib.AutoMapper.ValueTypeProfile)
            .AddAutoMapper(m => m.AddProfiles(typeCache, loggerFactory))
            .AddSingleton<ShoppingServiceMock>()
            .AddAlias<IShoppingService, ShoppingServiceMock>(ServiceLifetime.Singleton);

        var provider = services.BuildServiceProvider();
        _provider = provider;
        _shoppingService = provider.GetRequiredService<ShoppingServiceMock>();
        _idRegistry = provider.GetRequiredService<IIdRegistry>();
        _disposables.Add(provider);
        exceptions.ThrowIfNotEmpty();
    }
    [Fact]
    public void Version1_ManualDataNaive()
    {
        CustomerId customerId = new(Guid.NewGuid());
        _shoppingService.AddCustomer(
            new CustomerEntity("Customer")
            {
                Id = customerId
            }
        );
        OrderId orderId = new(Guid.NewGuid());
        _shoppingService.AddOrders(
            new OrderEntity(customerId, OrderStatus.Cart)
            {
                Id = orderId
            }
        );
        _shoppingService.SetCart(customerId, orderId);
        ArticleId firstArticle = new(Guid.NewGuid());
        ArticleId secondArticle = new(Guid.NewGuid());
        ArticleId thirdArticle = new(Guid.NewGuid());
        ArticleId fourthArticle = new(Guid.NewGuid());
        _shoppingService.AddArticles(
            new(nameof(firstArticle), 10)
            {
                Id = firstArticle
            },
            new(nameof(secondArticle), 20.20)
            {
                Id = secondArticle
            },
            new(nameof(thirdArticle), 30)
            {
                Id = thirdArticle
            },
            new(nameof(fourthArticle), 40.40)
            {
                Id = fourthArticle
            }
        );
        OrderItemId firstOrderItemId = new(Guid.NewGuid());
        OrderItemId secondOrderItemId = new(Guid.NewGuid());
        _shoppingService.AddOrderItems(
                new OrderItemEntity(orderId, firstArticle, 2, 10)
                {
                    Id = firstOrderItemId
                },
                new OrderItemEntity(orderId, secondArticle, 2, 20)
                {
                    Id = secondOrderItemId
                }
            );

        _shoppingService.AddArticleToCart(customerId, thirdArticle, 10);

        Action testFails = () => _shoppingService.MatchSnapshot();
        testFails.Should().Throw<EqualException>();
    }
    [Fact]
    public void Version1_ManualData()
    {
        CustomerId customerId = new(Guid.Parse("00000001-0000-0000-0000-000000000000"));
        _shoppingService.AddCustomer(
            new CustomerEntity("Customer")
            {
                Id = customerId
            }
        );
        OrderId orderId = new(Guid.Parse("00000002-0000-0000-0000-000000000000"));
        _shoppingService.AddOrders(
            new OrderEntity(customerId, OrderStatus.Cart)
            {
                Id = orderId
            }
        );
        _shoppingService.SetCart(customerId, orderId);
        ArticleId firstArticle = new(Guid.Parse("00000003-0000-0000-0000-000000000000"));
        ArticleId secondArticle = new(Guid.Parse("00000003-0001-0000-0000-000000000000"));
        ArticleId thirdArticle = new(Guid.Parse("00000003-0002-0000-0000-000000000000"));
        ArticleId fourthArticle = new(Guid.Parse("00000003-0003-0000-0000-000000000000"));
        _shoppingService.AddArticles(
            new(nameof(firstArticle), 10)
            {
                Id = firstArticle
            },
            new(nameof(secondArticle), 20.20)
            {
                Id = secondArticle
            },
            new(nameof(thirdArticle), 30)
            {
                Id = thirdArticle
            },
            new(nameof(fourthArticle), 40.40)
            {
                Id = fourthArticle
            }
        );
        OrderItemId firstOrderItemId = new(Guid.Parse("00000004-0000-0000-0000-000000000000"));
        OrderItemId secondOrderItemId = new(Guid.Parse("00000004-0001-0000-0000-000000000000"));
        _shoppingService.AddOrderItems(
                new OrderItemEntity(orderId, firstArticle, 2, 10)
                {
                    Id = firstOrderItemId
                },
                new OrderItemEntity(orderId, secondArticle, 2, 20)
                {
                    Id = secondOrderItemId
                }
            );

        _shoppingService.AddArticleToCart(customerId, thirdArticle, 10);

        _shoppingService.MatchSnapshot();
    }
    /// <summary>
    /// 
    /// </summary>
    [Fact]
    public void Version2_NaiveWay()
    {
        _provider.IncludeDataPackages<TwoOrderItemsDp<CustomerCartDp<CustomerDp>>>();

        _provider.GetRequiredServices(out CustomerDp customer, out ArticleDp articles);
        _shoppingService.AddArticleToCart(customer.Customer, articles.ThirdArticle, 10);

        _shoppingService.MatchSnapshot();
    }
    [Fact]
    public void Version3_UsingSnapshotInfos()
    {
        _provider.IncludeDataPackages<TwoOrderItemsDp<CustomerCartDp<CustomerDp>>>();

        _provider.GetRequiredServices(out CustomerDp customer, out ArticleDp articles);
        _shoppingService.AddArticleToCart(customer.Customer, articles.ThirdArticle, 10);

        new ShoppingServiceSnapshotInfo(_shoppingService, _idRegistry).MatchSnapshot();
    }
    [Fact]
    public void Version4_AddingBeforeAfterCompare()
    {
        _provider.IncludeDataPackages<TwoOrderItemsDp<CustomerCartDp<CustomerDp>>>();

        _provider.GetRequiredServices(out CustomerDp customer, out ArticleDp articles);

        new ShoppingServiceSnapshotInfo(_shoppingService, _idRegistry).MatchSnapshot(new SnapshotNameExtension("setup"));

        _shoppingService.AddArticleToCart(customer.Customer, articles.ThirdArticle, 10);

        new ShoppingServiceSnapshotInfo(_shoppingService, _idRegistry).MatchSnapshot(new SnapshotNameExtension("result"));
    }

    public void Dispose()
    {
        foreach (var disposable in _disposables)
            disposable.Dispose();
    }
}