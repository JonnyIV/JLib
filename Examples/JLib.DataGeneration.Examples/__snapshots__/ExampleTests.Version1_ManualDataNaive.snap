{
  "Customers": {
    "CustomerId { Value = 84da78ba-5d65-4597-8083-3eefe27c674b }": {
      "Id": {
        "Value": "84da78ba-5d65-4597-8083-3eefe27c674b"
      },
      "UserName": "Customer"
    }
  },
  "Orders": {
    "OrderId { Value = da22a1bf-3a45-48a6-a5bf-60a48b0e9be2 }": {
      "Id": {
        "Value": "da22a1bf-3a45-48a6-a5bf-60a48b0e9be2"
      },
      "Customer": {
        "Value": "84da78ba-5d65-4597-8083-3eefe27c674b"
      },
      "Status": "Cart"
    }
  },
  "OrderItems": {
    "OrderItemId { Value = f54a26ce-9379-4315-ada2-cee75368821a }": {
      "Id": {
        "Value": "f54a26ce-9379-4315-ada2-cee75368821a"
      },
      "OrderId": {
        "Value": "da22a1bf-3a45-48a6-a5bf-60a48b0e9be2"
      },
      "ArticleId": {
        "Value": "b891ecb1-f8bf-497e-863f-b01e94ea7bee"
      },
      "Quantity": 2,
      "Price": 10.0
    },
    "OrderItemId { Value = a4ec7867-b16b-47b6-b2e5-f77ddfb447d1 }": {
      "Id": {
        "Value": "a4ec7867-b16b-47b6-b2e5-f77ddfb447d1"
      },
      "OrderId": {
        "Value": "da22a1bf-3a45-48a6-a5bf-60a48b0e9be2"
      },
      "ArticleId": {
        "Value": "d5669f72-0f88-40bf-b974-7572f0e2d62e"
      },
      "Quantity": 2,
      "Price": 20.0
    },
    "OrderItemId { Value = c85a6c33-44e7-4634-8284-e637f3ef640b }": {
      "Id": {
        "Value": "c85a6c33-44e7-4634-8284-e637f3ef640b"
      },
      "OrderId": {
        "Value": "da22a1bf-3a45-48a6-a5bf-60a48b0e9be2"
      },
      "ArticleId": {
        "Value": "46f0eaa9-f354-46eb-b04f-a9804fd862d5"
      },
      "Quantity": 10,
      "Price": 30.0
    }
  },
  "Articles": {
    "ArticleId { Value = b891ecb1-f8bf-497e-863f-b01e94ea7bee }": {
      "Id": {
        "Value": "b891ecb1-f8bf-497e-863f-b01e94ea7bee"
      },
      "Name": "firstArticle",
      "Price": 10.0
    },
    "ArticleId { Value = d5669f72-0f88-40bf-b974-7572f0e2d62e }": {
      "Id": {
        "Value": "d5669f72-0f88-40bf-b974-7572f0e2d62e"
      },
      "Name": "secondArticle",
      "Price": 20.2
    },
    "ArticleId { Value = 46f0eaa9-f354-46eb-b04f-a9804fd862d5 }": {
      "Id": {
        "Value": "46f0eaa9-f354-46eb-b04f-a9804fd862d5"
      },
      "Name": "thirdArticle",
      "Price": 30.0
    },
    "ArticleId { Value = d8220558-73ac-4f0c-877a-ad268fc601c2 }": {
      "Id": {
        "Value": "d8220558-73ac-4f0c-877a-ad268fc601c2"
      },
      "Name": "fourthArticle",
      "Price": 40.4
    }
  },
  "Carts": {
    "CustomerId { Value = 84da78ba-5d65-4597-8083-3eefe27c674b }": {
      "Value": "da22a1bf-3a45-48a6-a5bf-60a48b0e9be2"
    }
  }
}
